### PR TITLE
Convert char to windows unicode

### PR DESCRIPTION
**Pull request by josediazfer**
Not merge yet.  I think that should support only Windows XP or greatest versions for the time. what do you think about it? @xarkes @XVilka

### CODE REVIEW

File: libr/debug/p/native/maps/windows.c [hunk 1]
```diff
@@ -93,10 +93,14 @@ static RList *w32_dbg_modules(RDebug *dbg) {
 		goto err_w32_dbg_modules;
 	}
 	do {
+		char *mod_name;
+
 		ut64 baddr = (ut64)(size_t)me32.modBaseAddr;
-                mr = r_debug_map_new (me32.szModule, baddr, baddr + me32.modBaseSize, 0, 0);
+		mod_name = r_sys_conv_w32_to_char (me32.szModule);
+                mr = r_debug_map_new (mod_name, baddr, baddr + me32.modBaseSize, 0, 0);
```
Comment (XVilka) [inline libr/debug/p/native/maps/windows.c]: broken indentation.

File: libr/debug/p/native/maps/windows.c [hunk 2]
```diff
@@ -93,10 +93,14 @@ static RList *w32_dbg_modules(RDebug *dbg) {
...
+		free (mod_name);
                 if (mr) {
```
Comment (XVilka) [inline libr/debug/p/native/maps/windows.c]: And here - the whole block should be re-indented.

File: libr/debug/p/native/reg.c [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // included from p/debug_native.c
 // splitted for better reading/cleaning up
 
-static const char *r_debug_native_reg_profile(RDebug *dbg) {
+static char *r_debug_native_reg_profile(RDebug *dbg) {
```
Comment (XVilka) [inline libr/debug/p/native/reg.c]: why removing 'const' here?

File: libr/debug/p/native/w32.c [hunk 4]
```diff
@@ -433,81 +429,70 @@ static int debug_exception_event (DEBUG_EVENT *de) {
...
-		UnmapViewOfFile (map);
-		CloseHandle (handle_file_map);
-		return NULL;
+	/* Create a file mapping to get the file name. */
+	map = MapViewOfFile (handle_file_map, FILE_MAP_READ, 0, 0, 1);
+	if (!map || !w32_GetMappedFileName (GetCurrentProcess (), map, filename, MAX_PATH)) {
+		goto err_get_file_name_from_handle;
 	}
-
+	TCHAR temp_buffer[512];
```
Comment (XVilka) [inline libr/debug/p/native/w32.c]: Why TCHAR here then if it's Unicode-only?

File: libr/include/r_util/r_sys.h [hunk 5]
```diff
@@ -45,8 +45,23 @@ R_API char *r_sys_getdir(void);
 R_API int r_sys_chdir(const char *s);
 R_API int r_sys_cmd_str_full(const char *cmd, const char *input, char **output, int *len, char **sterr);
 #if __WINDOWS__
+#if UNICODE
+#define W32_TCHAR_FSTR "%S"
+#define W32_TCALL(name) name"W"
+#define r_sys_conv_char_to_w32(buf) r_str_mb_to_wc (buf)
```
Comment (XVilka) [inline libr/include/r_util/r_sys.h]: may be name them "r_sys_conv_utf8_to_utf16" instead?

File: libr/util/file.c [hunk 6]
```diff
@@ -865,44 +902,63 @@ R_API char *r_file_temp (const char *prefix) {
...
 
-	if (GetTempPathA (sizeof (tmpdir), tmpdir) == 0) {
+	if ((len = GetTempPath (MAX_PATH + 1, tmpdir)) == 0) {
 		path = r_sys_getenv ("TEMP");
 		if (!path) {
 			path = strdup ("C:\\WINDOWS\\Temp\\");
 		}
 	} else {
-		DWORD (WINAPI *glpn)(LPCSTR, LPCSTR, DWORD) = r_lib_dl_sym (GetModuleHandle (TEXT ("kernel32.dll")), "GetLongPathNameA");
+		tmpdir[len] = 0;
```
Comment (XVilka) [inline libr/util/file.c]: please don't use VLA - they allocate on stack. Create buffer in heap instead please.

File: libr/util/sys.c [hunk 7]
```diff
@@ -374,13 +379,16 @@ R_API int r_sys_crash_handler(const char *cmd) {
 
 R_API char *r_sys_getenv(const char *key) {
 #if __WINDOWS__ && !__CYGWIN__
-	static char envbuf[4096];
+	TCHAR envbuf[4096];
```
Comment (XVilka) [inline libr/util/sys.c]: May be make it global variable and nullify before?

File: libr/util/sys.c [hunk 8]
```diff
@@ -653,14 +661,20 @@ R_API char *r_sys_cmd_str(const char *cmd, const char *input, int *len) {
...
 	}
 #if __WINDOWS__ && !defined(__CYGWIN__)
-	return CreateDirectoryA (dir, NULL) != 0;
+	LPTSTR dir_ = r_sys_conv_char_to_w32 (dir);
+
+	ret = CreateDirectory (dir_, NULL) != 0;
+	free (dir_);
 #else
-	return mkdir (dir, 0755) != -1;
+	ret= mkdir (dir, 0755) != -1;
```
Comment (XVilka) [inline libr/util/sys.c]: missing space

**Comment by Maijin** [comment]
@josediazfer  use `sys/indent.sh`

File: libr/debug/p/native/reg.c (hunk 3 above)
Comment (josediazfer) [reply libr/debug/p/native/reg.c]: supress warning casting const char * to char*:
....
char* (*reg_profile) (RDebug *dbg);
} RDebugPlugin;

File: libr/debug/p/native/w32.c (hunk 4 above)
Comment (josediazfer) [reply libr/debug/p/native/w32.c]: "TCHAR *" aka "LPTSTR". TCHAR type can be WCHAR or CHAR this depends on the define 
UNICODE.  Currently UNICODE always is defined in the platform.mk file.

**Comment by radare** [comment]
ok to merge? or not yet?

**Comment by XVilka** [comment]
@josediazfer please rename those utf8-utf16 conversion functions and it's good to merge.

**Comment by XVilka** [comment]
Please update the PR with this renaming and we'll merge.

**Comment by josediazfer** [comment]
OK, I will make this renaming it...but i don´t agree

**Comment by XVilka** [comment]
Thank you!

**Pull request merged**